manpages/tomrc.man
.na
.nh
.TH tomrc 1 2.00 BL3DFS
.SH NAME
tomrc \- Make a mrc file from rgb image files.
.SH SYNOPSIS
tomrc [options] <input files...> <output image>
.SH DESCRIPTION
tomrc takes a series of rgb files and converts them
into an mrc stack.  The default x and y dimensions
are the x and y size of the first image.  Images
that are larger will be clipped, and images that
are smaller will be centered by a black border.
.SS Options
.TP
.B -x <width>
Change the default width of the output image.
.TP
.B -y <height>
Change the default height of the output image.
.TP
.B -c
Make the output file a color file.
The default is to convert and color images to black
and white using the rule gray level =
(red * 0.3) + (green * 0.59) + (blue * 0.11).
.SH AUTHOR
Jim Kremer 
.SH SEE ALSO
frommrc(1)
.SH BUGS
The color output files are not standard MRC files.
They are created with a mode of 16 and store the color
information in three unsigned bytes.

Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/03: Modified PIP to read newly invented "autodoc" files to determine
program options and generate manual pages sections.

10/6/03: IMOD_3-0-22
10/6/03: Fixed the default number of views value for transferfid; it was not
being initialized in etomo.

10/6/03: Reordered thread control to eliminate race condition that often 
prevented matchvol1 from running right after any preceeding com script.

10/2/03: IMOD_3-0-21
10/2/03: Modified Rotatevol, Matchvol, and Warpvol to work on volumes up to 
16K in any dimension; Rotatevol was failing for an output X/Y size of 4K by 4K.

10/1/03: Made imodinfo work properly for counting scattered points on subsets
of the volume.

10/1/03: Added message for opening bead fixer window.

10/1/03: Added ability to turn a plugin into an internal module that is added
to the Special menu, to provide Bead fixer on Windows.

9/29/03: Restructured some 3dmod include files to restore ability to build
plugins with minimal header files.

9/25/03: Made sure windows stayed on the screen when restoring Zap and Info
window positions.

9/25/03: Changed Midas to number sections from 1 instead of 0.

9/24/03: IMOD_3-0-20
9/24/03: Resolved problems with the Fortran graphics package so it works
on all systems, at least briefly.

9/24/03: Fixed some problems with event sending under Windows, added a
retry capability to make it work reliably on the Mac, and found that it
did not work with the multi-threaded Qt library on the SGI.

9/24/03: Used different window properties for saving and restoring window
positions, which fixed problems in Redhat 7.3 and Irix.  Made it not restore
Info window size since that is highly font-dependent.

9/23/03: Put Fortran graphics programs inside application directories on the
Mac, which made them work correctly, with the window raisable and resizable.

9/22/03: Removed a large unused common from smoothsurf, which was making it
fail under Cygwin.

9/19/03: Made sample.com files produce top.rec, etc, directly as the output
of Tilt to avoid replacing existing .rec files.

9/19/03: Made 3dmod require only a 15-bit RGB visual, not 16 bit.

9/18/03: IMOD_3-0-19
9/18/03: Added the ability to float (match contrast between sections) based
on a subarea, either the image displayed or the area inside of a rubberband
in the active Zap window.  Also added an autocontrast button to the Info
window with target contrasts settable from the options dialog.

9/17/03: Made the Zap window pan by one display pixel per pixel of mouse move
for zooms less than one.

9/16/03: Made 3dmod save and restore sizes and positions of the Zap and Info
windows, with different settings saved for different image sizes.  Zoom is
adjusted to fit the whole image to the window.  If there are no application
saved settings, the Zap window now opens up wit a size that fits a zoomed-down
image if appropriate.

9/15/03: Eliminated physical flipping of data in 3dmod to avoid very bad
performance when the image data gets into swap space.

9/15/03: Optimized image drawing for zooms equal to the inverse of an integer.

9/13/03: IMOD_3-0-18
9/12/03: Prevented buffer overruns in an array that holds the filename in
3dmod, fixed problems with model checksums that were leading to queries to
save an unchanged model, arranged the startup page in two columns, fixed a
problem with zooming up under Windows, and added a workaround to a display
problem with Qt 3.2.1 on the Mac.

9/11/03: IMOD_3-0-17  beta release with Etomo

9/9/03: IMOD_3-0-16
9/5/03: Added a program, Resamplemod, for rotating a microtubule model to be
perpendicular to the Z axis and creating points at each section or plane
along that axis.

9/4/03: Incorporated changes in Mtlengths to correct for oblique microtubules
in tilted sections and implemented PIP input.

8/29/03: Set up and documented installation procedures on the Mac.

8/29/03: Changed the Fortran graphics routines to run the graphics window in
a separate thread from the Fortran program.  This only works properly under
Linux, where the window refreshes itself when resized or exposed.  On the Mac,
the window can now be moved but not resized.

8/25/03: Added the -E option to Imodmesh to cap the ends of tubes.

8/13/03: IMOD_3-0-15
8/12/03: Made the Qt-based graphics routines used in Fortran programs work on
the Mac without hanging when the window opened.

8/12/03: Made it possible to start up Nda without a model file, just like Mtk.

8/9/03: IMOD_3-0-14
8/9/03: Changed Blendmont to take PIP input and modified command files and
Copytomocoms to use new form of input.

8/8/03: Etomo was incorporated into IMOD build structure.

8/8/03: Wrote man page for Mtk, changed program to make terminology
consistent, to fix problems with saving connector objects and moving items to
new objects, and to add ability to export a graph in a format easily readable
by other programs.  (Nda can also export graphs.)  Also provided some example
command files in the com directory.

8/8/03: Fixed problem of current point not being displayed when model is
turned off, caused by the ability to display the vector from bead fixer.

8/1/03: Made it possible to kill 3dmod during image flipping operation.

8/1/03: Standardized error exit in Xfproduct.

8/1/03: Implemented proper thickness padding in fast backprojection code of
Tilt, reduced Tilt's stack size so the increase is only 3-fold from previously,
and made memory usage and data loading dynamic so that it will not use more
than 80 MB of stack unless it needs to.

7/31/03: IMOD_3-0-13
7/31/03: Added ability to send messages to set movie mode and to load a model
without changing black/white levels.

7/31/03: Changed the call used to iconify dialog boxes along with the info
or model view window because the old way didn't work under RedHat 9 or on the 
SGI under Qt 3.1.

7/31/03: Fixed remaining problems created by new treatment of "views", and

mrc/tomrc.c
/*  IMOD VERSION 2.02
 *
 *  tomrc.c -- Convert SGI rgb files to an mrc file.
 *
 *  Author: James Kremer email: kremer@colorado.edu
 */

/*****************************************************************************
 *   Copyright (C) 1995-1996 by Boulder Laboratory for 3-Dimensional Fine    *
 *   Structure ("BL3DFS") and the Regents of the University of Colorado.     *
 *                                                                           *
 *   BL3DFS reserves the exclusive rights of preparing derivative works,     *
 *   distributing copies for sale, lease or lending and displaying this      *
 *   software and documentation.                                             *
 *   Users may reproduce the software and documentation as long as the       *
 *   copyright notice and other notices are preserved.                       *
 *   Neither the software nor the documentation may be distributed for       *
 *   profit, either in original form or in derivative works.                 *
 *                                                                           *
 *   THIS SOFTWARE AND/OR DOCUMENTATION IS PROVIDED WITH NO WARRANTY,        *
 *   EXPRESS OR IMPLIED, INCLUDING, WITHOUT LIMITATION, WARRANTY OF          *
 *   MERCHANTABILITY AND WARRANTY OF FITNESS FOR A PARTICULAR PURPOSE.       *
 *                                                                           *
 *   This work is supported by NIH biotechnology grant #RR00592,             *
 *   for the Boulder Laboratory for 3-Dimensional Fine Structure.            *
 *   University of Colorado, MCDB Box 347, Boulder, CO 80309                 *
 *****************************************************************************/

/* To compile:  cc tomrc.c -o readimage -limage */

#include <stdio.h>
#include <stdlib.h>

#ifndef __sgi
main(int argc, char **argv)
{
     fprintf(stderr, "%s only runs on sgi machines.\n", argv[0]);
}
#else

#include <gl/image.h>
#include "mrcfiles.h"

main(argc,argv)
     int argc;
     char **argv;
{
     struct MRCheader hdata;
     FILE *fout;
     IMAGE *image = NULL;
     int i;
     int x, y, z;
     int xsize = 0, ysize = 0, zsize;
     int ox, oy, oxsize, oysize;
     int xborder, yborder;
     int color = FALSE;
     short *rbuf, *gbuf, *bbuf;
     float avg = 0.0;



     if (argc < 3){
	  fprintf(stderr, "tomrc version 1.0 %s, %s\n",__DATE__,__TIME__);
	  fprintf(stderr, "Copyright (C)1994 Boulder Laboratory for\n");
	  fprintf(stderr,
		  "3-Dimensional Fine Structure, University of Colorado.\n"); 
	  fprintf(stderr, "Converts SGI rgb or bw files to mrc format.\n");
	  fprintf(stderr, "Usage: tomrc [-c] [-x #] [-y #] [infiles.rgb...] [outfile.mrc]\n");
	  fprintf(stderr, "Options:\n");
	  fprintf(stderr, "\t-c will cause a color mrc file to be made.\n");
	  fprintf(stderr, "\t-x xsize of output image. (Default, size of first input image.)\n");
	  fprintf(stderr, "\t-y ysize of output image.\n");
	  exit(1);
     }
     
     for (i = 1; i < argc; i++){
	  if (argv[i][0] == '-'){
	       switch (argv[i][1]){
		  case 'c':
		    color = TRUE;
		    break;

		  case 'x':
		    sscanf(argv[++i], "%d", &xsize);
		    break;

		  case 'y':
		    sscanf(argv[++i], "%d", &ysize);
		    break;

		  default:
		    break;

	       }
	  }else 
	       break;
     }

/*     printf("Opening %s\n", argv[i]); */
     if( (image=iopen(argv[i],"r")) == NULL ) {
	  fprintf(stderr,"tomrc: can't open input file %s\n",argv[i]);
	  exit(1);
     }

     
     fout = fopen(argv[argc - 1], "w");
     
     if (fout == NULL){
	  fprintf(stderr,"tomrc: Couldn't open output file %s\n", argv[2]);
	  exit(1);
     }
     
     if (xsize <= 0)
	  xsize = image->xsize;
     if (ysize <= 0)
	  ysize = image->ysize;

     zsize = argc - i - 1;
     

     mrc_head_new( &hdata, xsize, ysize, zsize, 0);
     hdata.amin = image->min;
     hdata.amax = image->max;
     hdata.amean = 128;

     if ((color)  && (image->zsize >= 3))
	  hdata.mode = MRC_MODE_RGB;

     mrc_head_write(fout, &hdata);

     iclose(image);

/*     printf("xsize,ysize = %d,%d\n",xsize,ysize); */
     while (i < (argc - 1)){
	  if( (image = iopen(argv[i],"r")) == NULL ) {
	       fprintf(stderr,"tomrc: can't open input file %s\n",argv[i]);
	       exit(1);
	  }
/*	  printf("Opened %s, %d,%d\n", argv[i],image->xsize,image->ysize); */

	  rbuf = (short *)malloc(image->xsize * sizeof(short));
	  gbuf = (short *)malloc(image->xsize * sizeof(short));
	  bbuf = (short *)malloc(image->xsize * sizeof(short));

	  xborder = xsize - image->xsize;
	  yborder = ysize - image->ysize;
	  xborder /= 2;
	  yborder /= 2;

/*	  printf("xyborder = (%d, %d)\n", xborder, yborder);  */

	  oysize = image->ysize;

	  /* fill bottom border */
	  if (yborder >= 0){
	       for(y = 0; y < yborder; y++)
		    for(x = 0; x < xsize; x++){
			 fputc(0x00,fout);
			 if (color){
			      fputc(0x00,fout);
			      fputc(0x00,fout);
			 }
		    }
	       oy = 0;
	  }else{
	       oy = -1 * yborder;
	       oysize = ysize + oy;
	  }

	  for(; oy < oysize; oy++, y++){
/*	       printf("row %d, line %d\n", oy, y); */
	       getrow(image,rbuf,oy,0);
	       if (image->zsize > 2){
		    getrow(image,gbuf,oy,1);
		    getrow(image,bbuf,oy,2);
	       }
	       
	       /* fill left border */
	       if (xborder >= 0){
		    for(x = 0; x < xborder; x++){
			 fputc(0x00,fout);
			 if (color){
			      fputc(0x00,fout);
			      fputc(0x00,fout);
			 }
		    }
		    ox = 0;
		    oxsize = image->xsize;
		    
	       }else{
		    x = -xborder;
		    oxsize = xsize + x;
		    ox = x;

	       }
	       
	       /* fill image */
	       for (; ox < oxsize; ox++, x++){
		    if (color){
			 
			 fputc(rbuf[ox],fout);
			 fputc(gbuf[ox],fout);
			 fputc(bbuf[ox],fout);
			 
		    }
		    if (image->zsize > 2){
			 rbuf[ox] = ((float)rbuf[ox] * 0.3) + 
			      ((float)gbuf[ox] * 0.59) + 
				   ((float)bbuf[ox] * 0.11);
		    }

		    if (!color)
			 fputc(rbuf[ox], fout);

		    avg += rbuf[x];
	       }

	       /* fill right border */
	       for (;x<xsize;x++){
		    fputc(0x00,fout);
		    if (color){
			 fputc(0x00,fout);
			 fputc(0x00,fout);
			      }
	       }
	  }

	  /* fill bottom border */
	  for(;y<ysize;y++)
	       for(x=0;x<xsize;x++){
		    fputc(0x00,fout);
		    if (color){
			 fputc(0x00,fout);
			 fputc(0x00,fout);
		    }
	       }

	  free(rbuf);
	  free(gbuf);
	  free(bbuf);
	  iclose(image);
	  i++;
     }

     avg /= xsize * ysize * zsize;
     hdata.amean = avg;
     mrc_head_write(fout, &hdata);
     
     fclose(fout);
     
}


#endif
